busylib/client.py

```diff
@@ -1,108 +1,315 @@
+import dataclasses
+import enum
+import json
+
 import requests
-from typing import IO
-from busylib.types import ApiResponse
-
-
-class ApiClient:
-    def __init__(self, base_url: str):
-        if not base_url.endswith("/"):
-            base_url += "/"
-        self.base_url = base_url
-        self.session = requests.Session()
-
-    def _request(
-        self,
-        method: str,
-        endpoint: str,
-        params: dict[str, object] | None = None,
-        json: object | None = None,
-        data: object | None = None,
-        headers: dict[str, str] | None = None,
-    ) -> ApiResponse:
-        url = f"{self.base_url}{endpoint}"
+
+from busylib import exceptions
+from busylib import types
+
+
+def _serialize_for_json(obj):
+    if isinstance(obj, enum.Enum):
+        return obj.value
+    elif isinstance(obj, dict):
+        return {k: _serialize_for_json(v) for k, v in obj.items()}
+    elif isinstance(obj, list):
+        return [_serialize_for_json(item) for item in obj]
+    elif dataclasses.is_dataclass(obj):
+        return _serialize_for_json(dataclasses.asdict(obj))
+    else:
+        return obj
+
+
+class BusyBar:
+    """
+    Main library class for interacting with the Busy Bar API.
+    """
+
+    def __init__(self, addr: str):
+        self.base_url = f"http://{addr}"
+        self.client = requests.Session()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.close()
+
+    def close(self):
+        self.client.close()
+
+    def _handle_response(self, response: requests.Response, as_bytes: bool = False):
+        if response.status_code >= 400:
+            try:
+                error_data = response.json()
+                raise exceptions.BusyBarAPIError(
+                    error=error_data.get("error", "Unknown error"),
+                    code=error_data.get("code", response.status_code),
+                )
+            except json.JSONDecodeError:
+                raise exceptions.BusyBarAPIError(
+                    error=f"HTTP {response.status_code}: {response.text}",
+                    code=response.status_code,
+                )
```

Comment:
mb add raise_error? idk if it's a good idea or not to raise an error every >= 400 request